bash.md

```diff
@@ -0,0 +1,116 @@
+## Bash
+
+### Introduction
+
+First, if you haven't already, see `linux-basics.md` for a primer on Linux. Fundamentally, all commands take in the following input:
+
+- Path of the executable to run
+- Arguments to the executable
+- Data from file descriptors
+
+And all commands will output:
+- Data to file descriptors
+- A 32bit integer return code
+
+Of course, they can also open files, access the network, etc. But we won't count that as input/output from the operating system's point of view. That's just what the program is doing to your computer.
+
+#### Arguments and return code
+
+```c
+#include <stdio.h>
+int main(int argc, char* argv[]) {
+  printf("Argc: %d", argc);
+  for(int i = 0; i < argc; i++) {
+    printf("Argv[%d]: %s", i, argv[i]);
+  }
+  return 0;
+}
+```
+
+This executable, will print out argc, all of argv, and then give a return code of 0. Return codes of 0 are expected to mean that a function _succeeded_, if you want to express failure, you can `return -1;` or some other non-zero value.
+
+#### File Descriptors
+
+What is a file descriptor? Very easy. A file descriptor is a device on Linux that you can either send data to, or read data from. Every command has 1024 of them, labelled from 0 to 1023. Most of them remain unused though and thus `unbound` since it's not connected to any device. File descriptors can be bound to your terminal window, so you can see what the output of a command is or send input to the command via typing. Or, they can be bound to files, so that commands can directly write their output to a file. By default, a command is expected to _read_ from file descriptor 0, _write_ to file descriptor 1, and write _errors_ to file descriptor 2. In libc, you're guaranteed to get the following `#define`'s in `stdio.h`,
```

Comment:
I'm concerned about the phrasing here. Since we describe the differences between file descriptors and streams later, and explicitly show file descriptors to be integers, can we use the definition of them being references to open file objects, instead of a device on Linux? I worry this might confuse astute students since 'device' has special meaning in Linux